File: dotnet/UpsChecklist.Core/ChecklistPositions.cs
using System.Text.Json;
using UpsChecklist.Core.Models;

namespace UpsChecklist.Core;

public static class ChecklistPositions
{
    private static readonly Lazy<IReadOnlyList<Position>> Positions = new(BuildPositions);

    public static IReadOnlyList<Position> All => Positions.Value;

    public static Position? Get(string id) => All.FirstOrDefault(p => p.Id == id);

    public static IReadOnlyList<CheckItem> AllItems(Position position) =>
        position.Sections.SelectMany(s => s.Items).ToList();

    public static string? ScheduleWarning(Position position, string date)
    {
        if (string.IsNullOrEmpty(date) || position.Schedule is null)
            return null;

        var mondaySelected = DateTime.TryParse(date, out var parsed) && parsed.DayOfWeek == DayOfWeek.Monday;
        if (position.Schedule == "monday" && !mondaySelected)
            return "This combined PS1 / PS2 checklist is marked Monday only in the workbook. Check your selected date or assignment.";
        if (position.Schedule == "other-days" && mondaySelected)
        {
            return position.Id == "pd2"
                ? "PD2 is marked Not on Monday in the workbook. Confirm your assignment with your team leader."
                : "For Monday, the workbook provides the combined PS1 / PS2 checklist. Check your selected position.";
        }

        return null;
    }

    private static IReadOnlyList<Position> BuildPositions()
    {
        var sourcePath = Path.Combine(AppContext.BaseDirectory, "Data", "workbook-source.json");
        var source = JsonSerializer.Deserialize<WorkbookSource>(
            File.ReadAllText(sourcePath),
            new JsonSerializerOptions { PropertyNameCaseInsensitive = true })
            ?? throw new InvalidOperationException("Missing workbook source.");

        string Cell(string sheet, string address)
        {
            var found = source.Sheets.FirstOrDefault(s => s.Name == sheet)
                ?? throw new InvalidOperationException($"Missing source sheet: {sheet}");
            if (!found.Cells.TryGetValue(address, out var value) || string.IsNullOrWhiteSpace(value))
                throw new InvalidOperationException($"Missing source instruction: {sheet}!{address}");
            return value.Trim();
        }

        CheckItem Item(string sheet, string address, string? noteAddress = null) =>
            new(address.ToLowerInvariant(), Cell(sheet, address), noteAddress is null ? null : Cell(sheet, noteAddress), address);

        ReferenceImage Ref(string file, string label) => new(file, label);

        CheckSection Before(string sheet) => new(
            "before",
            "Before the sort",
            [
                Item(sheet, "B8"),
                Item(sheet, "B10"),
                Item(sheet, "B12", sheet == "Ps1-Ps2 monday only" ? "F12" : null),
            ]);

        (string id, string label, string title, string sourceSheet) Base(string sheet, string id, string label) =>
            (id, label, Cell(sheet, "E3"), sheet);

        Position Ps(string sheet, string id, string label, string image, int firstRow)
        {
            var b = Base(sheet, id, label);
            return new Position(
                b.id,
                b.label,
                b.title,
                b.sourceSheet,
                [
                    Before(sheet),
                    new CheckSection(
                        "after",
                        "After the sort",
                        [
                            Item(sheet, $"B{firstRow}"),
                            Item(sheet, $"B{firstRow + 2}"),
                            Item(sheet, $"B{firstRow + 4}"),
                        ],
                        [Ref(image, $"{label} conveyor diagram")]),
                ],
                Cell(sheet, $"B{firstRow + 6}"),
                Schedule: "other-days",
                ScheduleLabel: "Other days");
        }

        Position Pd(string sheet, string id, string label, string image)
        {
            var finish = id == "pd3" ? 37 : 36;
            var b = Base(sheet, id, label);
            return new Position(
                b.id,
                b.label,
                b.title,
                b.sourceSheet,
                [
                    Before(sheet),
                    new CheckSection(
                        "after",
                        "After the sort",
                        [Item(sheet, "B25"), Item(sheet, "F25")],
                        [Ref(image, $"{label} conveyor diagram")]),
                    new CheckSection(
                        "floor",
                        "Work floor",
                        [Item(sheet, "B27")],
                        [Ref("image4.png", "First part of the metro belt")]),
                    new CheckSection(
                        "finish",
                        "Finish the checks",
                        [Item(sheet, $"B{finish}"), Item(sheet, $"B{finish + 2}")]),
                ],
                Cell(sheet, "G29"),
                Schedule: id == "pd2" ? "other-days" : null,
                ScheduleLabel: id == "pd2" ? "Not on Monday" : null);
        }

        const string monday = "Ps1-Ps2 monday only";
        const string smalls = "Smalls";
        const string matrix = "Matrix";

        var mondayBase = Base(monday, "ps1-ps2-monday", "PS1 / PS2");
        var psMonday = new Position(
            mondayBase.id,
            mondayBase.label,
            mondayBase.title,
            mondayBase.sourceSheet,
            [
                Before(monday),
                new CheckSection(
                    "ps1",
                    "After the sort: PS1",
                    [Item(monday, "B20"), Item(monday, "B22")],
                    [Ref("image2.png", "PS1 conveyor diagram")]),
                new CheckSection(
                    "ps2",
                    "After the sort: PS2",
                    [Item(monday, "B30"), Item(monday, "B32")],
                    [Ref("image3.png", "PS2 conveyor diagram")]),
                new CheckSection(
                    "finish",
                    "Finish the checks",
                    [Item(monday, "B35", "F35")],
                    Note: "Confirm timing: the source labels this after-sort check ‘When you start’."),
            ],
            Cell(monday, "B37"),
            Schedule: "monday",
            ScheduleLabel: "Monday only",
            SourceNotes:
            [
                "The Monday sheet says \"When you start\" beside its after-sort clean-position check (B35/F35). This wording is preserved; confirm the intended timing with your team leader.",
            ]);

        var smallsBase = Base(smalls, "smalls", "Smalls");
        var smallsPosition = new Position(
            smallsBase.id,
            smallsBase.label,
            smallsBase.title,
            smallsBase.sourceSheet,
            [
                Before(smalls),
                new CheckSection(
                    "sls1",
                    "After the sort: SLS1",
                    [
                        Item(smalls, "B30"), Item(smalls, "F30"), Item(smalls, "B34"),
                        Item(smalls, "B36"), Item(smalls, "B38"), Item(smalls, "B40"),
                    ],
                    [Ref("image9.png", "SLS1 marked conveyor diagram")]),
                new CheckSection(
                    "sls1-under",
                    "SLS1: under and around",
                    [Item(smalls, "A32")],
                    [Ref("image11.png", "SLS1 marked area to check under and around")],
                    RemarksKey: "sls1"),
                new CheckSection(
                    "sls2",
                    "After the sort: SLS2",
                    new List<CheckItem>(new[] { 68, 70, 72, 74, 76, 78, 80 }.Select(row => Item(smalls, $"B{row}"))),
                    [Ref("image10.png", "SLS2 marked conveyor diagram")]),
            ],
            Cell(smalls, "B82"),
            Handover: Cell(smalls, "B90"),
            SourceNotes:
            [
                "The source places the walk-off instruction under Before the sort, although it explicitly says to check after the sort. The original wording is preserved.",
                "The paper handover instruction says to sign the first page, but the signature boxes are on the second page. Confirm the digital handover process; downloading does not send the report.",
            ]);

        var matrixBase = Base(matrix, "matrix", "Matrix");
        var matrixPosition = new Position(
            matrixBase.id,
            matrixBase.label,
            matrixBase.title,
            matrixBase.sourceSheet,
            [
                Before(matrix),
                new CheckSection(
                    "recirculation",
                    "After the sort: recirculation",
                    [
                        Item(matrix, "B30"), Item(matrix, "B32", "C32"),
                        Item(matrix, "B34", "C34"), Item(matrix, "B36", "C36"),
                    ],
                    [Ref("image12.png", "Matrix recirculation diagram")],
                    RemarksKey: "recirculation"),
                new CheckSection(
                    "da",
                    "DA platform",
                    new List<CheckItem>(new[] { 53, 55, 57, 59 }.Select(row => Item(matrix, $"B{row}"))),
                    [Ref("image13.png", "DA platform belt diagram")]),
                new CheckSection(
                    "chutes",
                    "Chutes and collector belts",
                    [Item(matrix, "B74"), Item(matrix, "B76"), Item(matrix, "B83")],
                    [Ref("image14.png", "Matrix chute reference photo")]),
            ],
            Cell(matrix, "E84"));

        return
        [
            psMonday,
            Ps("Ps1 other days", "ps1", "PS1", "image2.png", 20),
            Ps("Ps2 other days", "ps2", "PS2", "image3.png", 21),
            Pd("PD1", "pd1", "PD1", "image5.png"),
            Pd("PD2 Not on Monday", "pd2", "PD2", "image6.png"),
            Pd("PD3", "pd3", "PD3", "image7.png"),
            Pd("PD4", "pd4", "PD4", "image8.png"),
            smallsPosition,
            matrixPosition,
        ];
    }
}

File: lib/checklist-positions.ts
import source from "./workbook-source.json";

export type CheckItem = { id: string; text: string; note?: string; sourceCell: string };
export type ReferenceImage = { file: string; label: string };
export type CheckSection = { id: string; title: string; items: CheckItem[]; images?: ReferenceImage[]; note?: string; remarksKey?: string };
export type Position = { id: string; label: string; title: string; sourceSheet: string; schedule?: "monday" | "other-days"; scheduleLabel?: string; sections: CheckSection[]; packageLabel: string; handover?: string; sourceNotes?: string[] };
export type PositionDraft = { checks: Record<string, boolean>; count: string; remarks: string; sectionRemarks: Record<string, string>; evidencePhoto: string; signature: string; drawn: string; signatureMode: "type" | "draw" };
export const emptyDraft = (): PositionDraft => ({ checks: {}, count: "", remarks: "", sectionRemarks: {}, evidencePhoto: "", signature: "", drawn: "", signatureMode: "type" });

const sheets = source.sheets;
function sheet(name: string) {
  const found = sheets.find(s => s.name === name);
  if (!found) throw new Error(`Missing source sheet: ${name}`);
  return found;
}
function cell(name: string, address: string) {
  const value = (sheet(name).cells as Record<string, string | undefined>)[address];
  if (!value) throw new Error(`Missing source instruction: ${name}!${address}`);
  return value.trim();
}
function item(name: string, address: string, noteAddress?: string): CheckItem {
  return { id: address.toLowerCase(), text: cell(name, address), sourceCell: address, ...(noteAddress ? { note: cell(name, noteAddress) } : {}) };
}
const ref = (file: string, label: string): ReferenceImage => ({ file, label });
function before(name: string): CheckSection {
  return { id: "before", title: "Before the sort", items: [item(name, "B8"), item(name, "B10"), item(name, "B12", name === "Ps1-Ps2 monday only" ? "F12" : undefined)] };
}
function base(name: string, id: string, label: string) {
  return { id, label, title: cell(name, "E3"), sourceSheet: name };
}
function ps(name: string, id: string, label: string, image: string, firstRow: number): Position {
  return { ...base(name, id, label), schedule: "other-days", scheduleLabel: "Other days", packageLabel: cell(name, `B${firstRow + 6}`), sections: [before(name), { id: "after", title: "After the sort", images: [ref(image, `${label} conveyor diagram`)], items: [item(name, `B${firstRow}`), item(name, `B${firstRow + 2}`), item(name, `B${firstRow + 4}`)] }] };
}
function pd(name: string, id: string, label: string, image: string): Position {
  const finish = id === "pd3" ? 37 : 36;
  return { ...base(name, id, label), ...(id === "pd2" ? { schedule: "other-days" as const, scheduleLabel: "Not on Monday" } : {}), packageLabel: cell(name, "G29"), sections: [before(name), { id: "after", title: "After the sort", images: [ref(image, `${label} conveyor diagram`)], items: [item(name, "B25"), item(name, "F25")] }, { id: "floor", title: "Work floor", images: [ref("image4.png", "First part of the metro belt")], items: [item(name, "B27")] }, { id: "finish", title: "Finish the checks", items: [item(name, `B${finish}`), item(name, `B${finish + 2}`)] }] };
}
const monday = "Ps1-Ps2 monday only";
const smalls = "Smalls";
const matrix = "Matrix";

export const positions: Position[] = [
  { ...base(monday, "ps1-ps2-monday", "PS1 / PS2"), schedule: "monday", scheduleLabel: "Monday only", packageLabel: cell(monday, "B37"), sourceNotes: ['The Monday sheet says "When you start" beside its after-sort clean-position check (B35/F35). This wording is preserved; confirm the intended timing with your team leader.'], sections: [before(monday), { id: "ps1", title: "After the sort: PS1", images: [ref("image2.png", "PS1 conveyor diagram")], items: [item(monday, "B20"), item(monday, "B22")] }, { id: "ps2", title: "After the sort: PS2", images: [ref("image3.png", "PS2 conveyor diagram")], items: [item(monday, "B30"), item(monday, "B32")] }, { id: "finish", title: "Finish the checks", note: "Confirm timing: the source labels this after-sort check ‘When you start’.", items: [item(monday, "B35", "F35")] }] },
  ps("Ps1 other days", "ps1", "PS1", "image2.png", 20),
  ps("Ps2 other days", "ps2", "PS2", "image3.png", 21),
  pd("PD1", "pd1", "PD1", "image5.png"),
  pd("PD2 Not on Monday", "pd2", "PD2", "image6.png"),
  pd("PD3", "pd3", "PD3", "image7.png"),
  pd("PD4", "pd4", "PD4", "image8.png"),
  { ...base(smalls, "smalls", "Smalls"), packageLabel: cell(smalls, "B82"), handover: cell(smalls, "B90"), sourceNotes: ["The source places the walk-off instruction under Before the sort, although it explicitly says to check after the sort. The original wording is preserved.", "The paper handover instruction says to sign the first page, but the signature boxes are on the second page. Confirm the digital handover process; downloading does not send the report."], sections: [before(smalls), { id: "sls1", title: "After the sort: SLS1", images: [ref("image9.png", "SLS1 marked conveyor diagram")], items: [item(smalls, "B30"), item(smalls, "F30"), item(smalls, "B34"), item(smalls, "B36"), item(smalls, "B38"), item(smalls, "B40")] }, { id: "sls1-under", title: "SLS1: under and around", images: [ref("image11.png", "SLS1 marked area to check under and around")], items: [item(smalls, "A32")], remarksKey: "sls1" }, { id: "sls2", title: "After the sort: SLS2", images: [ref("image10.png", "SLS2 marked conveyor diagram")], items: [68, 70, 72, 74, 76, 78, 80].map(row => item(smalls, `B${row}`)) }] },
  { ...base(matrix, "matrix", "Matrix"), packageLabel: cell(matrix, "E84"), sections: [before(matrix), { id: "recirculation", title: "After the sort: recirculation", images: [ref("image12.png", "Matrix recirculation diagram")], items: [item(matrix, "B30"), item(matrix, "B32", "C32"), item(matrix, "B34", "C34"), item(matrix, "B36", "C36")], remarksKey: "recirculation" }, { id: "da", title: "DA platform", images: [ref("image13.png", "DA platform belt diagram")], items: [53, 55, 57, 59].map(row => item(matrix, `B${row}`)) }, { id: "chutes", title: "Chutes and collector belts", images: [ref("image14.png", "Matrix chute reference photo")], items: [item(matrix, "B74"), item(matrix, "B76"), item(matrix, "B83")] }] },
];

export const getPosition = (id: string) => positions.find(p => p.id === id);
export const allItems = (position: Position) => position.sections.flatMap(s => s.items);
export function scheduleWarning(position: Position, date: string): string | undefined {
  if (!date || !position.schedule) return;
  const mondaySelected = new Date(`${date}T12:00:00Z`).getUTCDay() === 1;
  if (position.schedule === "monday" && !mondaySelected) return "This combined PS1 / PS2 checklist is marked Monday only in the workbook. Check your selected date or assignment.";
  if (position.schedule === "other-days" && mondaySelected) return position.id === "pd2" ? "PD2 is marked Not on Monday in the workbook. Confirm your assignment with your team leader." : "For Monday, the workbook provides the combined PS1 / PS2 checklist. Check your selected position.";
}
